Задача 2

ASHGOLDOFFICIAL opened the pull request:
# Changelog

## Общее по проекту

- Подключение логирования (#2, #17). 
- Добавлены GitHub Actions с двумя простыми задачами, которые запускаются на Ubuntu 24.04 LTS при создании PR (в том числе этого) (#9):
    - Пакетирование и запуск тестов (запускает `mvn package` и `mvn test`).
    - Проверка стиля кода.

## Платформы

- Добавлена возможность ядром отправлять кнопки (реализована на платформах Discord и Telegram) (#11):
- Добавлен общий интерфейс для ботов (#7)

## Хранение данных

- Описаны модели хранимых данных (#10, #12):
    - `User` хранит пользователя нашего бота.
    - `UserEntry` хранит зарегистрированные пользователем данные (отдельньная модель, потому что предполагается, что у одного пользователя может быть несколько таковых).
- Написан интерфейс хранилища (см. паттерн "Репозиторий") и его реализация на массиве (#10, #12). Используется для хранения объектов, описанных моделями.

## Логическое ядро

- Ядро теперь не возвращает ботам сообщения для отправки, вместо этого оно напрямую вызывает у них метод отправки сообщения (#10).
- Написано ядро основного бота по паттерну "Состояние" (#12), что позволяет вести с пользователем многошаговый диалог (без классов со сложными ветвлениями).
- Реализован процесс регистрации и удаления данных (#12). Данные хранятся в выше описанных хранилищах.
- Реализован процесс редактирования данных (#18).

# Работа бота

Работа бота происходит по следующей схеме:

1. Пользователь отправляет сообщение.
1. Бот некоторой платформы его получает и отправляет ядру.
1. Ядро проверяет, есть ли данный пользователь в хранилище (условно БД).
    1. Если нет, добавляет этого пользователя в БД и присваевает ему дефолтное состояние.
1. Ядро достаёт из БД этого пользователя.
1. Ядро берёт объект, обрабатывающий сообщения для текущего состояния пользователя (этот объект -- состояние логического ядра или мини-ядро).
1. Ядро передаёт сообщение пользователя этому состоянию.
1. Состояние обрабатывает внутри сообщение: валидизирует полученные данные, сохраняет/удаляет данные в/из БД, при необходимости меняет состояние пользователя на некоторое другое.
1. Схема повторяется.

Таким образом, основное ядро выступает роутером между отдельными состояниями логического ядра (мини-ядрами). Это позволяет писать обработчики сообщений для конкретного состояния пользователя в отдельном классе сравнительно независимо друг от друга. Более подробно про работу кода читайте в #12 и комментариях к нему.

DJettt committed:
Добавление кнопок в сообщении в TelegramBot (пока не завершено)
+ Поменял название класса Message -> LocalMessage всвязи с тем, что и в тг боте и в дс боте существует класс Message
+ Добавил проверку на наличие CallBack запроса в сообщении.
- Пока не добавил саму генерацию кнопки при отправке сообщения.

DJettt committed:
Добавление кнопок в Телеграм Бота.

LocalButton: Добавление файла
* Класс содержит в себе поле, характеризующее все кнопки, которые нужно добавить в сообщение бота.

LocalMessage: Была переделана структура файла
* Теперь в сообщении между ядром и ботами будет передаваться характеристика сообщения (они разные для разных сторон) в переменной MESSAGE_STATUS

* В случае обнаружения неизвестного статуса LOGGER выводит ошибку в терминал.

* Так же было добавлено поле buttons содержащие в себе информацию о кнопках, которые должны присутствовать в сообщении.

Примеры статусов:
Core -> Bot
* "text_only" - в сообщении 'бот -> пользователь' не содержится информация о кнопках (они не планируются).
* "text_with_buttons" - в сообщении 'бот ->пользователь' должны быть кнопки. Информация о них содержится в поле LocalMessage.button.
TelegramBot -> Core
* "callback_query" - пользователь нажал на кнопку в боте.
* "message" - пользователь отправил текстовое сообщение боту.

EchoBotCore: Переделана структура.
* Ядро теперь может создавать кнопки и отправлять их (пока что только для одного сообщения и кнопки не генерируются, так как метод создавался только для проверки)
* При отправке сообщения ядро устанавливает его характеристику.
* При получении LocalMessage ядро проверяет сначала статус сообщения, потом уже выполняет нужные действия и вызывает методы.

TelegramBot: Переделана структура для работы с кнопками.
* Добавлены методы:
- createButton: создает объект одной кнопки
- createButtons: создает всю сетку кнопок.
- sendBackStatusCallback: устанавливает нужный статус
- sendBackStatusMessage: устанавливает нужный статус

* Переделаны методы:
- consume: определяет какой тип обновления произошел
- createFromMessage: теперь обрабатывает статус сообщения от ядра

* Методы, которые теперь не используются:
- createFromTelegramMessage: метод не был удален

DJettt committed:
Bug fix
Корректная работа вспомогательных кнопок, созданных специально для проверки их работоспособности.

DJettt committed:
Добавление кнопок в DiscordBot
+ Добавление кнопок в бота Discord. Кнопки полностью функционируют, как и в тг, но:
- кнопки могут идти только в строку (не сетка как в Телеграме)
- после нажатия на кнопку бот отправляет нужную информацию, но под кнопками вылезает ошибка "This interaction failed".
Чуть подробнее про новые методы:
+ onButtonInteraction - отслеживает нажатия на кнопки.
+ createButton - создает кнопку Дискорда из объекта LocalButton
+ createButtons - создает контейнер кнопок из переданной в бота сетки кнопок
LocalMessage:
+ Изменен конструктор по строке. Сейчас при создании объекта в нем присваивается статус "string_only".
TelegramBot - незначительные изменения
EchoBotCore - незначительные изменения

DJettt committed:
Merge branch 'main' into Add-buttons

# Conflicts:
#	src/main/java/ru/urfu/Bot.java
#	src/main/java/ru/urfu/DiscordBot.java
#	src/main/java/ru/urfu/EchoBotCore.java
#	src/main/java/ru/urfu/Message.java
#	src/main/java/ru/urfu/TelegramBot.java

DJettt committed:
Разрешение конфликтов с веткой main.

ASHGOLDOFFICIAL committed:
Изменение архитектуры: избавление от статусов сообщений.

DJettt committed:
Изменения оформления кода с помощью плагина checkstyle

DJettt committed:
Корректировка оформления кода с помощью плагина checkstyle
+ Подкорректированы JavaDoc-и
+ Подкорректировано оформление некоторых участков кода.

DJettt committed:
Merge branch 'Add-buttons-changes' into Add-buttons

# Conflicts:
#	src/main/java/ru/urfu/DiscordBot.java
#	src/main/java/ru/urfu/EchoBotCore.java
#	src/main/java/ru/urfu/LocalButton.java
#	src/main/java/ru/urfu/LocalMessage.java
#	src/main/java/ru/urfu/TelegramBot.java

ASHGOLDOFFICIAL committed:
Перенёс логические ядра в отдельный подпакет, превратил LogicCore в интерфейс.

ASHGOLDOFFICIAL committed:
Превратил Message в рекорду, добавил ему билдер, поправил тесты.

ASHGOLDOFFICIAL committed:
Методы ботов вызываются из ядра.

ASHGOLDOFFICIAL committed:
Набросал методы-пустышки для логики основного бота.

ASHGOLDOFFICIAL committed:
Правил стиль, перенёс тест в нужную папку.

ASHGOLDOFFICIAL committed:
Создал простой интерфейс хранилищ (Storage), написал реализацию на массиве, добавил тестов.

ASHGOLDOFFICIAL committed:
Добавил модели объектов, которые будут храниться и использоваться в основном боте.

ASHGOLDOFFICIAL committed:
Кинул ботов в отдельную папку.

ASHGOLDOFFICIAL committed:
Правки стиля.

DJettt committed:
Корректный вывод кнопок в обоих ботах.
В Дискорде ограничение кнопок - 5, поэтому, если нужно отправить больше кнопок, то бот будет делать это в отдельных сообщениях.
В Телеграме максимальное количество кнопок в ряду - 8, поэтому вывод кнопок будет частично напоминать квадрат.

ASHGOLDOFFICIAL committed:
Merge pull request #10 from DJettt/task-2-shaat-logic

База для написания основного бота

DJettt committed:
Merge branch 'task-2' into Add-buttons

# Conflicts:
#	src/main/java/ru/urfu/LogicCore.java
#	src/main/java/ru/urfu/Message.java
#	src/main/java/ru/urfu/bots/DiscordBot.java
#	src/main/java/ru/urfu/bots/TelegramBot.java
#	src/main/java/ru/urfu/logics/EchoBotCore.java
#	src/test/java/ru/urfu/logics/EchoBotCoreTest.java

ASHGOLDOFFICIAL committed:
Добавил хранилище на хэш-мапе.

DJettt committed:
Решенный конфликт между ветками task-2 и Add-buttons.
Ветка Add-buttons готова к слиянию с веткой task-2

DJettt committed:
Правки стиля.

DJettt committed:
Merge pull request #11 from DJettt/Add-buttons

Add buttons

ASHGOLDOFFICIAL committed:
Merge remote-tracking branch 'refs/remotes/origin/task-2' into task-2-shaat-logic

ASHGOLDOFFICIAL committed:
Расширил функционал хранилищ.

ASHGOLDOFFICIAL committed:
Введение объектов многошаговых процессов и их состояний.

DJettt committed:
Рабочий код с удалением данных о себе.

ASHGOLDOFFICIAL committed:
Примитивный /delete.

DJettt committed:
Merge branch 'refs/heads/task-2-shaat-logic' into task-2-storage-interactions-pribylov

# Conflicts:
#	src/main/java/ru/urfu/logics/MathMechBotCore.java
#	src/main/java/ru/urfu/models/User.java
#	src/main/java/ru/urfu/storages/Storage.java

ASHGOLDOFFICIAL committed:
Удалил метод builder у LocalMessage (ибо без static он бесполезен), немного изменил обработчик команд.

DJettt committed:
Добавил infoCommandHandler

ASHGOLDOFFICIAL committed:
Merge remote-tracking branch 'refs/remotes/origin/task-2-storage-interactions-pribylov' into task-2-shaat-logic

# Conflicts:
#	src/main/java/ru/urfu/logics/MathMechBotCore.java

DJettt committed:
Добавление регистрации пользователя.

ASHGOLDOFFICIAL committed:
Merge remote-tracking branch 'refs/remotes/origin/task-2-storage-interactions-pribylov' into task-2-shaat-logic

ASHGOLDOFFICIAL committed:
Добавил состояний процессу редактирования информации.

DJettt committed:
Изменение оформления кода.

ASHGOLDOFFICIAL committed:
Изменил структуру проекта и ядра в частности. Теперь используем паттерн "State" для ядра.

ASHGOLDOFFICIAL committed:
Вынес кнопки "Да", "Нет", "Назад" в константы.

ASHGOLDOFFICIAL committed:
Тесты для команды удаления.

ASHGOLDOFFICIAL committed:
Первые тесты для команды регистрации, небольшие правки стиля.

ASHGOLDOFFICIAL committed:
Разбиение состояния запроса ФИО в процессе регистрации на методы, добавлен тест.

DJettt committed:
Разделили регистрацию на 2 смысловые части - обработка команды и отправка информации.

ASHGOLDOFFICIAL committed:
Разбиение состояния запроса года обучения в процессе регистрации на методы, добавлены тесты.

ASHGOLDOFFICIAL committed:
Превратил State в интерфейс.

ASHGOLDOFFICIAL committed:
Ядро изменяет состояние при получении сообщения от пользователя (теперь можно корректно вести диалог с несколькими пользователями).

ASHGOLDOFFICIAL committed:
Тесты для состояний запроса направления во время регистрации, вынес логику из RegistrationSpecialitiesState#processMessage в отдельные методы.

ASHGOLDOFFICIAL committed:
Подправил оформление тестов. Добавил в уместные тесты случай пустой строки.

ASHGOLDOFFICIAL committed:
Правки стиля.

ASHGOLDOFFICIAL committed:
Сделал модели вновь рекордами, добавил хранилища с методами изменения данных, поправил стиль.

ASHGOLDOFFICIAL committed:
Сделал модели вновь рекордами, добавил хранилища с методами изменения данных, поправил стиль, добавил тест.

ASHGOLDOFFICIAL committed:
Перенёс всё, что касается ММ-бота в отдельную папку; теперь enum'ы состояний хранят класс состояния, enum'ы процесса - классы enum'ов состояний (на будущее).

ASHGOLDOFFICIAL committed:
Теперь отдельные состояния не могут менять состояния всего контекста. Убрал лишние поля из модели User. Enum StateList теперь содержит список состояний, каждое из которых содержит соответсвующий класс: теперь для получения объекта состояния достаточно просто иметь пользователя с этим состоянием в качестве поля.

ASHGOLDOFFICIAL committed:
Удалил бесполезные классы списков процессов: списки состояний делают всю работу.

ASHGOLDOFFICIAL committed:
Процесс удаления теперь не удаляет пользователя, только введённые им данные.

ASHGOLDOFFICIAL committed:
Поменял модификаторы классов и методов, ассерты, стиль.

ASHGOLDOFFICIAL committed:
Выделил константы в отдельный класс, поменял тип State с абстрактного класса на интерфейс.

ASHGOLDOFFICIAL committed:
Стейты в отдельной папке, добавил структуру, состоящую из двух хранилищ (можно воспринимать как БД, где хранилища - таблицы).

ASHGOLDOFFICIAL committed:
Ренеймы: StateList -> State, State -> LogicCoreState (чтобы указать на связь стейтов с ядром).

ASHGOLDOFFICIAL committed:
Ренеймы: State -> UserState (и наследуемые классы по той же схеме). LogicCoreState теперь расширяет LogicCore.

ASHGOLDOFFICIAL committed:
Убрал не используемый конструктор, добавил доки.

ASHGOLDOFFICIAL committed:
Вместо метода LogicCoreState#onEnter(LocalMessage, long, Bot), который отправлял сообщение, теперь LogicCoreState#enterMessage(long), который просто возвращает сообщение для отправки. В теории этот метод может быть статичным.

ASHGOLDOFFICIAL committed:
Все стейты логического ядра теперь синглтон-енамы; удалил лишние, крайне абстрактные интерфейсы.

ASHGOLDOFFICIAL committed:
Правки стиля.

ASHGOLDOFFICIAL committed:
Правки стиля.

ASHGOLDOFFICIAL committed:
Раскидал по папкам.

ASHGOLDOFFICIAL committed:
Вынес создание storage из ядра, чтобы затем во время тестов можно было мокировать его.

ASHGOLDOFFICIAL committed:
Удалил HashMapStorage (не использовался), изменил сигнатуры методов Storage (теперь принимают сущности, а не их id), адаптировал код под изменения.

ASHGOLDOFFICIAL committed:
Попытался максимально добавить @NotNull/@Nullable аннотации на публичные методы, чистка кода.

ASHGOLDOFFICIAL committed:
Обновил тесты для состояния запроса ФИО (регистрация), исправил валидацию ФИО: теперь можно использовать букву `ё` и разделять ФИО любым количеством whitespace символов.

ASHGOLDOFFICIAL committed:
Переписал тесты для состояния запроса года обучения (регистрация). Небольшие исправления по пути.

ASHGOLDOFFICIAL committed:
Изменил реализацию метода ArrayStorage#delete на более простую. Не должно повлиять на работу остального кода.

ASHGOLDOFFICIAL committed:
Обновил тесты для состояния запроса направления подготовки (регистрация).

ASHGOLDOFFICIAL committed:
Переместил файл Identifiable.java в корректное место.

ASHGOLDOFFICIAL committed:
Переместил билдеры моделей в папку к самим моделям. Немного приукрасил код.

ASHGOLDOFFICIAL committed:
Обновил тесты для удаления.

ASHGOLDOFFICIAL committed:
Обновил тесты и исправил некорректное поведение в RegistrationGroupState.

ASHGOLDOFFICIAL committed:
Обновил тесты и исправил некорректное поведение в RegistrationGroupState.

ASHGOLDOFFICIAL committed:
Все Optional#get() -> Optional#orElseThrow() в тестах. Стиль.

ASHGOLDOFFICIAL committed:
Обернул частые аргументы функций LocalMessage msg, long id, Bot bot в одну структуру: Request.

ASHGOLDOFFICIAL committed:
Объединил два состояния запроса специальности в одно (минус два класса).

ASHGOLDOFFICIAL committed:
Стиль.

ASHGOLDOFFICIAL committed:
small: Объединил две строки в одну.

DJettt committed:
Merge remote-tracking branch 'origin/task-2-shaat-logic' into task-2-storage-interactions-pribylov

# Conflicts:
#	src/main/java/ru/urfu/enums/processes/EditionProcessState.java
#	src/main/java/ru/urfu/enums/processes/Process.java
#	src/main/java/ru/urfu/enums/processes/ProcessState.java
#	src/main/java/ru/urfu/logics/MathMechBotCore.java
#	src/main/java/ru/urfu/logics/mathmechbot/models/Specialty.java
#	src/main/java/ru/urfu/models/User.java
#	src/main/java/ru/urfu/models/UserEntry.java
#	src/main/java/ru/urfu/storages/ArrayStorage.java

ASHGOLDOFFICIAL committed:
Исправил проблему с некорректным выводом информации о человеке без отчества. Добавил соответсвующий тест. Вынес создание строки информацию в метод модели UserEntry.

ASHGOLDOFFICIAL committed:
Сделал тест более хранилища универсальным.

ASHGOLDOFFICIAL committed:
Разделил файл с тестами матмех бота на несколько файлов, каждый из которых тестирует отдельные стейты. Пока что нет красивой вложенности.

ASHGOLDOFFICIAL committed:
Сделал MathMechStorage final.

ASHGOLDOFFICIAL committed:
Плюс тесты.

ASHGOLDOFFICIAL committed:
Подключил проверку стиля в тестах. Подправил стиль.

ASHGOLDOFFICIAL committed:
Убрал логирование пользователя-отправителя в главном ядре.

ASHGOLDOFFICIAL committed:
Убрал комменты из pom.xml (уже понятно всё, я думаю).

DJettt committed:
Добавил команду /edit, но пока что можно изменить только ФИО

DJettt reviewed (approved):
Все классно 👍

ASHGOLDOFFICIAL committed:
Merge pull request #12 from DJettt/task-2-shaat-logic

Логическое ядро нового бота

DJettt committed:
Добавил возможность изменить свой курс.

ASHGOLDOFFICIAL commented:
> Все классно 👍

Дальше будет лучше.

ASHGOLDOFFICIAL committed:
Обновил конфигурацию логирования. Теперь для нашего пакета доступны все уровни логов.

ASHGOLDOFFICIAL committed:
Merge pull request #17 from DJettt/task-2-logging-shaat

Настройка логирования

DJettt committed:
Добавил возможность изменить свое направление.

DJettt committed:
Добавил возможность изменить свою группу и строку МЕН.

DJettt committed:
Корректировка оформления.

DJettt committed:
Корректировка оформления + bugfix.

DJettt committed:
Merge branch 'refs/heads/task-2' into task-2-storage-interactions-pribylov

DJettt committed:
Чуть-чуть подкорректировал тесты, обновив /help

ASHGOLDOFFICIAL committed:
Инкапсулировал users и userEntries в MathMechStorage.

ASHGOLDOFFICIAL committed:
Merge pull request #19 from DJettt/task-2-public-fields-shaat

Инкапсуляция users и userEntries в MathMechStorage.

DJettt committed:
Инкапсуляция

DJettt committed:
Merge branch 'task-2' into task-2-storage-interactions-pribylov

# Conflicts:
#	src/main/java/ru/urfu/logics/mathmechbot/states/DefaultState.java
#	src/main/java/ru/urfu/logics/mathmechbot/storages/MathMechStorage.java

DJettt committed:
Добавил тесты для состояний изменения.

DJettt committed:
Новые тесты.

DJettt committed:
Bugfix

DJettt committed:
Bugfix

DJettt committed:
Merge pull request #18 from DJettt/task-2-storage-interactions-pribylov

Добавление команды /edit, которая дает возможность пользователю менять информацию о себе.

DJettt committed:
Инкапсуляция storage и небольшое дополнение в тестах для более точной проверки.

vpyzhyanov commented on src/main/java/ru/urfu/Main.java:
```diff
@@ -24,7 +31,7 @@ private Main() {}
     private static void startBot(LogicCore logicCore, String env, Class<? extends Bot> botClass) {
```
> Class<? extends Bot> botClass

Самое время познакомиться с полиморфизмом. Это намного проще, эффективнее и красивее. Сейчас тут оверинжиниринг, который даже непонятно зачем нужен. 

Сделайте так же как с LogicCore. Просто создайте нужный экземпляр и передайте его сюда. LogicCore надо убрать, он не нужен, он внутри бота

vpyzhyanov replied:
Вроде только начал смотреть и в самом первом методе уже куча проблем

![image](https://github.com/user-attachments/assets/16aa4a81-b7b1-4ab9-8cb6-f99bc2ea4ba7)

1. Проблема с расширяемостью кода. Код явно ориентирован не только на телеграмм
2. Уже написал, но повторюсь. Проблема с полиморфизмом и лишние аргументы
3. Нет никакой проблемы вызывать заранее известный вам конструктор. Тем более тут вы создали себе проблему в будущем. Класс бота обязан реализовывать указанный вами конструктор, но вы нигде это не контролируете. Да и не надо ничего это делать, в этом нет смысла.
4. Запуск бота не должен быть в интерфейсе Bot. Тут вы обязываете бота запускаться без параметров. Но почему вы решили наложить на него такие строгие требования? Не надо так. Смысла в этом тоже ноль.
5. Логирование - это очень хорошо, но использовать его надо правильно. Например в данном случае, вы осознанно решили, что нужно "глушить" ошибку и продолжать выполнение программы? Правда есть смысл работать дальше, если бот не запустился? Конечно же нет.

Мы плавно подошли к тому что метод не содержит ООП и на самом деле вообще не нужен. Особенно об этом кричит пункт 4, ибо название метода говорит, что только ради этого он и существует.

Правда ведь не стоило писать аж 20 строчек кода просто чтобы заменить этот код:

```java
        TelegramBot telegramBot = new TelegramBot(System.getenv("TGMATHMECHBOT_TOKEN"), logicCore);
        telegramBot.start();
```

Классы и методы нужны, чтобы облегчить сложные конструкции и упростить код, а не увеличить его.

ASHGOLDOFFICIAL replied:
8683a95fc27249233d9ef8979e6960451931bc6a

Буду прикладывать коммиты с решением проблемы, чтобы отслеживать, что уже решено.

vpyzhyanov replied:
лучше словами пишите что именно изменили, отдельные коммиты никто смотреть не будет

vpyzhyanov commented on src/main/java/ru/urfu/Main.java:
```diff
@@ -24,7 +31,7 @@ private Main() {}
     private static void startBot(LogicCore logicCore, String env, Class<? extends Bot> botClass) {
         String botToken = System.getenv(env);
```
кстати, неудачное имя переменной `env,` оно ни о чём не говорит. Например оно должно быть `botTokenEnvirontmentName`. 

Но правильно просто передавать **значение**, а не имя переменной

ASHGOLDOFFICIAL replied:
8683a95fc27249233d9ef8979e6960451931bc6a

vpyzhyanov commented on src/main/java/ru/urfu/Main.java:
```diff
@@ -41,8 +48,11 @@ private static void startBot(LogicCore logicCore, String env, Class<? extends Bo
      */
     public static void main(String[] args) {
-        final LogicCore logicCore = new EchoBotCore();
+        final MathMechStorage storage = new MathMechStorage(new UserArrayStorage(), new UserEntryArrayStorage());
```
абсолютно всегда тут используются одни и те же классы. Значит их можно не устанавливать через конструктор, а инициализировать внутри конструктора по умолчанию.

ASHGOLDOFFICIAL replied:
ac5b97a15c34ae277e6b9640ee9e5fc5342bf95b

vpyzhyanov commented on src/main/java/ru/urfu/bots/Bot.java:
```diff
@@ -1,7 +1,10 @@
+ * Интерфейс бота, описывающий методы, которые ожидаются в каждом боте.
  */
 public interface Bot {
     /**
```
Метод start надо удалить, он для бота не нужен

ASHGOLDOFFICIAL replied:
8683a95fc27249233d9ef8979e6960451931bc6a

vpyzhyanov commented on src/main/java/ru/urfu/Main.java:
```diff
@@ -24,7 +31,7 @@ private Main() {}
     private static void startBot(LogicCore logicCore, String env, Class<? extends Bot> botClass) {
```
**static запрещён!!!!**

ASHGOLDOFFICIAL replied:
8683a95fc27249233d9ef8979e6960451931bc6a

vpyzhyanov commented:
Задача гарантировано принята не будет

![image](https://github.com/user-attachments/assets/712e69e8-795e-4a2d-b00a-35909af1cfcc)

![image](https://github.com/user-attachments/assets/1e57eb04-f703-4d82-a13a-38d8204b8888)

vpyzhyanov commented on src/main/java/ru/urfu/bots/DiscordBot.java:
```diff
@@ -0,0 +1,186 @@
+ * в зависимости от переданного ему при создании логического ядра (logicCore).
+ */
+public final class DiscordBot extends ListenerAdapter implements Bot {
+    private static final Logger LOGGER = LoggerFactory.getLogger(DiscordBot.class);
```
static запрещён!!!!

ASHGOLDOFFICIAL replied:
169d35feaaacbbd1b6c524e36d428dfef39701ea

vpyzhyanov commented on src/main/java/ru/urfu/bots/DiscordBot.java:
```diff
@@ -0,0 +1,186 @@
+     */
+    @Override
+    public void start() {
+        //TODO: проверить на возникновение исключений
```
надо проверить

ASHGOLDOFFICIAL replied:
Оставляю Артёму.

DJettt replied:
Проверил, исключения не возникнут.

vpyzhyanov commented on src/main/java/ru/urfu/bots/DiscordBot.java:
```diff
@@ -0,0 +1,186 @@
+                        )
+                )
+                .setStatus(OnlineStatus.ONLINE)
+                .setActivity(Activity.playing("IDEA Intellij"))
```
каким образом бот связан со средой разработки?

ASHGOLDOFFICIAL replied:
e038826d61cbfd8fed3e674f831d98b0a3758105

ASHGOLDOFFICIAL replied:
Изначально подразумевалось как небольшой элемент юмора.

DJettt replied:
В данном случае мы устанавливали статус бота в самом Дискорд. После запуска программы около бота было написано "играет в IDEA Intellij". 
В любом случае мы убрали этот статус в исправлениях.

vpyzhyanov commented on src/main/java/ru/urfu/bots/DiscordBot.java:
```diff
@@ -0,0 +1,186 @@
+    private List<List<LocalButton>> splitButtons(LocalMessage message) {
+        assert message.buttons() != null;
+
+        final int maxSize = 5;
```
это должно быть константой

ASHGOLDOFFICIAL replied:
9bf39508ef1226009097078afcdd872e6d3315c0

vpyzhyanov commented on src/main/java/ru/urfu/bots/DiscordBot.java:
```diff
@@ -0,0 +1,186 @@
+     * @param id id чата куда нужно отправить сообщение.
+     */
+    @Override
+    public void sendMessage(@NotNull LocalMessage message, @NotNull Long id) {
```
Переопределять JavaDoc у интерфейсов категорически запрещено! Это надругательство над полиморфизмом, сводящее его в могилу

ASHGOLDOFFICIAL replied:
9bf39508ef1226009097078afcdd872e6d3315c0

vpyzhyanov commented on src/main/java/ru/urfu/bots/DiscordBot.java:
```diff
@@ -0,0 +1,186 @@
+    }
+
+    /**
+     * Отслеживает отправление сообщения от пользователя боту.
```
аналогично.

Если вам нужны свои комментарии, то напишите их в виде комментария внутри метода

ASHGOLDOFFICIAL replied:
9bf39508ef1226009097078afcdd872e6d3315c0

vpyzhyanov commented on src/main/java/ru/urfu/bots/TelegramBot.java:
```diff
@@ -0,0 +1,195 @@
+ * в зависимости от переданного ему при создании логического ядра (logicCore).
+ */
+public final class TelegramBot implements Bot, LongPollingSingleThreadUpdateConsumer {
+    private static final Logger LOGGER = LoggerFactory.getLogger(TelegramBot.class);
```
опять static!

ASHGOLDOFFICIAL replied:
169d35feaaacbbd1b6c524e36d428dfef39701ea

vpyzhyanov commented on src/main/java/ru/urfu/bots/TelegramBot.java:
```diff
@@ -0,0 +1,195 @@
+ * Простой телеграм-бот, который принимает текстовые сообщения и составляет ответ
+ * в зависимости от переданного ему при создании логического ядра (logicCore).
+ */
+public final class TelegramBot implements Bot, LongPollingSingleThreadUpdateConsumer {
```
я смотрю вы лёгких путей не ищете. Почему не наследовались от `TelegramLongPollingBot`?

ASHGOLDOFFICIAL replied:
Примеры из документации используют `LongPollingSingleThreadUpdateConsumer`. `TelegramLongPollingBot` обычно упоминается в Интернет-примерах. Было решено использовать то, что использует документация, а не люди со сторонних сайтов.

ASHGOLDOFFICIAL replied:
7fa4593e159c3a0c29937f5da70a5b6f8b9748ad (хэш коммита в https://github.com/rubenlagus/TelegramBots)

Вообще, есть подозрения, что `TelegramLongPollingBot` deprecated. Подтверждений найти не могу, но автор удалил папку с этим классом в коммите с хэшем, указанным выше. Возможно, он был куда-то перенесён, но я не могу найти куда. В любом случае, если в документации используют `LongPollingSingleThreadUpdateConsumer`, то думаю, что не просто так.

ASHGOLDOFFICIAL replied:
![image](https://github.com/user-attachments/assets/29236312-09eb-4671-85c0-2ec50fbb6b95)

https://rubenlagus.github.io/TelegramBotsDocumentation/how-to-update-7.html#migrating-your-existing-longpolling-bots

Нашёл подтверждение. `TelegramLongPollingBot` можно считать deprecated.

vpyzhyanov commented on src/main/java/ru/urfu/bots/TelegramBot.java:
```diff
@@ -0,0 +1,195 @@
+        telegramClient = new OkHttpTelegramClient(token);
+        logicCore = core;
+        botToken = token;
+        botsApplication = new TelegramBotsLongPollingApplication();
```
Из-за такого "хардкорного" решения теперь у вас бот внутри бота. Очень интересно обоснование зачем это было сделано, ну вдруг тут есть какой-то тайный смысл (хотя вряд ли)

ASHGOLDOFFICIAL replied:
`TelegramBotsLongPollingApplication` и `LongPollingSingleThreadUpdateConsumer` -- это два разных класса, у которых нет ни общих потомков, ни общих интерфейсов. Поэтому честно не вижу проблем. 

Если я правильно понял, что автор библиотеки хотел, то:
`LongPollingSingleThreadUpdateConsumer` -- это обработчик апдейтов от Telegram.
`TelegramBotsLongPollingApplication` -- это класс, отвечающий за регистрацию и запуск реализаций `LongPollingUpdateConsumer`.

Поэтому прошу уточнить, в чём проблема создания `TelegramBotsLongPollingApplication` внутри наследника `LongPollingUpdateConsumer`, если он используется для старта лишь одного единственного (этого же) `LongPollingUpdateConsumer`.

vpyzhyanov commented on src/main/java/ru/urfu/bots/TelegramBot.java:
```diff
@@ -0,0 +1,195 @@
+    }
+
+    /**
+     * Запуск бота в отдельном потоке.
```
тут вы совершили страшное преступление. Поменяли логику заданную интерфейсом. Полиморфизм теперь с вами не дружит

ASHGOLDOFFICIAL replied:
8cdd9817b35393d3bbc2bbd130e96fe795f8e455

vpyzhyanov commented on src/main/java/ru/urfu/bots/TelegramBot.java:
```diff
@@ -0,0 +1,195 @@
+     */
+    @Override
+    public void start() {
+        new Thread(() -> {
```
зачем новый поток? Если что, телеграмм бот и так запускается в отдельном потоке (по крайней мере в простом и красивом варианте)

ASHGOLDOFFICIAL replied:
Не запускается. Как минимум, в `LongPollingSingleThreadUpdateConsumer` он запускается в `main` потоке, не давая запуститься DiscordBot. Выбор  `LongPollingSingleThreadUpdateConsumer` был описан в соответвующем замечании.

vpyzhyanov commented on src/main/java/ru/urfu/bots/TelegramBot.java:
```diff
@@ -0,0 +1,195 @@
+
+                Thread.currentThread().join();
+            } catch (Exception e) {
+                LOGGER.error("Bot didn't get registered and didn't run.", e);
```
вы сознательно "глушите" ошибку? Какой смысл продолжать, если ничего не работает?

ASHGOLDOFFICIAL replied:
DiscordBot может запуститься и работать без TelegramBot. Если не запустятся оба, то программа закончит исполннение, не создав ни одного потока.

8cdd9817b35393d3bbc2bbd130e96fe795f8e455

vpyzhyanov replied:
Так себе аргумент. Действительно, половина приложения не запустилась, ну и ладно, пусть пользователи пользуются тем, что запустилось сегодня, может завтра другое запуститься.......

vpyzhyanov commented on src/main/java/ru/urfu/bots/TelegramBot.java:
```diff
@@ -0,0 +1,195 @@
+    }
+
+    /**
+     * Отправление сообщения, формат которого содержится в msg.
```
аналогично

ASHGOLDOFFICIAL replied:
8cdd9817b35393d3bbc2bbd130e96fe795f8e455

vpyzhyanov commented on src/main/java/ru/urfu/bots/TelegramBot.java:
```diff
@@ -0,0 +1,195 @@
+            msg = convertTelegramMessage(update.getMessage());
+            chatId = update.getMessage().getChatId();
+        } else {
+            LOGGER.error("Unknown message type!");
```
нет никакой информации о том, что же там было. Вам на поддержке будет очень сложно понять что пошло не так. Добавьте дополнительную информацию, чтобы быстро понять что произошло. Например передайте туда `update`

ASHGOLDOFFICIAL replied:
8cdd9817b35393d3bbc2bbd130e96fe795f8e455

vpyzhyanov commented on src/main/java/ru/urfu/localobjects/LocalMessageBuilder.java:
```diff
@@ -0,0 +1,48 @@
+/**
+ * Билдер для класса LocalMessage.
+ */
+public final class LocalMessageBuilder {
```
как правило билдер помещают внутрь класса, который он создаёт, чтобы он умел обращаться к методам, к которым не могут обращаться другие, например к приватному конструктору.

ASHGOLDOFFICIAL replied:
В таком случае нужно делать весь внутренний класс static, либо делать статичный геттер билдера. Что-то вроде:

```java
public static LocalMessageBuilder builder() {
    return new LocalMessageBuilder();
}
```

В противном случае придётся создавать LocalMessage, чтобы создать LocalMessageBuilder, что убивает смысл билдера (ибо объект уже сделан). Вы на паре запретили использовать static для этих целей.

DJettt replied:
Я обсудил это с Вами на паре, Вы убедились, что без static тут не обойтись. В итоге было решено оставить билдер в другом классе.

vpyzhyanov commented on src/main/java/ru/urfu/localobjects/LocalMessageBuilder.java:
```diff
@@ -0,0 +1,48 @@
+/**
+ * Билдер для класса LocalMessage.
+ */
+public final class LocalMessageBuilder {
```
смысла в таком билдере мало, т.к. параметра всего 2 и можно легко создать необходимые варианты конструкторов

ASHGOLDOFFICIAL replied:
Параллельно ведутся попытки научить LocalMessage хранить картинки и прочие прикреплённые медиа-файлы. Билдер используется, чтобы потом весь код не переписывать, расширяя аргументы конструктора.

DJettt replied:
da0d034721e24b3d6b7997626c843556d8cf2f94
Был добавлен конструктор в LocalMessage и исправлены подобные моменты в коде.

vpyzhyanov commented on src/main/java/ru/urfu/localobjects/Request.java:
```diff
@@ -0,0 +1,18 @@
+ * @param bot     бот, принявший сообщение.
+ */
+public record Request(
+        long id,
```
интересно почему и в каком таком удивительном случае id может быть null. По идее не должен быть никогда

ASHGOLDOFFICIAL replied:
30a195ef9af888c47c5eb0fdd585a73848db06de

Полагаю, это может быть достигнуто путём приминения обёртки, но я не смог придумать как именно.

vpyzhyanov replied:
Вы не поняли, речь про аннотацию `@NotNull`, а идентификатор единственное поле, которое ей не помечено, значит оно может быть null

vpyzhyanov commented on src/main/java/ru/urfu/logics/LogicCore.java:
```diff
@@ -0,0 +1,17 @@
+     *
+     * @param request запрос.
+     */
+    void processMessage(@NotNull Request request);
```
с одной стороны удобно сгруппировать всё это в отдельный класс, но тут есть довольно существенные недостатки

1. Количество связей в программе увеличивается на единицу, что усложняет код
2. Параметров довольно мало (всего 3 штуки), что ставит под сомнение необходимость усложнения кода
3. **(очень критично)** В эту группу могут попасть параметры, которые никак сюда по смыслу не подходят, что и произошло. Но это замечание я напишут в Request

Единственный параметр, достойный выделения в отдельный класс - это Message, но это вы уже сделали.

ASHGOLDOFFICIAL replied:
Список аргументов становился слишком велик (в плане длины строчки). Желательно было у всех методов нынче принимающих Request оставлять одинаковые названия для переменных (для единообразия кода), приходилось вручную следить за этим. Поэтому было решено объединить эти аргументы, которые всегда ходили вместе, в один класс.

Пока не решаю проблему, подожду Артёма.

DJettt replied:
С точки зрения абстракции с этим не сложно работать и такой вариант реализации вполне простой, так как в запросе есть вся информация, которая нужна ядру при общении с пользователем.
id - идентификатор чата.
LocalMessage - само сообщение  bot->core.
Bot - информация о том, откуда пришло сообщение / куда отправить сообщение.

То есть request - запрос от бота к ядру
LocalMessage - содержимое запроса (сообщение от пользователя)
bot - отправитель запроса / адресат запроса

DJettt replied:
24e487069342e2adfe768c43ec1010354f1298a6
Убрана зависимость в виде объекта Request

vpyzhyanov commented on src/main/java/ru/urfu/localobjects/Request.java:
```diff
@@ -0,0 +1,18 @@
+public record Request(
+        long id,
+        @NotNull LocalMessage message,
+        @NotNull Bot bot
```
Класс называется "запрос". Но бот к запросу никакого отношения не имеет. Он занимается **отправкой** сообщения, а это не запрос, а **ответ**. Нарушена логика и читаемость кода сильно снижена.

Нужно убрать это поле из этого класса

ASHGOLDOFFICIAL replied:
Не понимаю, почему это ответ. Request -- это запрос бота ядру обработать полученное сообщение. Содержит минимум нужных для этого аргументов: само сообщение (чтобы понять что обрабатывать), id отправителя, bot принявший сообщение (нужен для того, чтобы в дальнейшем понимать, с какой платформы пришло сообщение. От этого как минимум трактовка id может отличаться). Поэтому bot тут нужен пока что для понимания, откуда сообщение пришло.

Жду Артёма.

DJettt replied:
Пока что решили оставить это название, поскольку так работа с ядром более понятна. Согласен с Андреем по поводу "Request -- это запрос бота ядру".

vpyzhyanov replied:
обсудили на паре и не нашли ничего удобного в классе Request

vpyzhyanov commented on src/main/java/ru/urfu/logics/mathmechbot/Constants.java:
```diff
@@ -0,0 +1,28 @@
+    public final static String BACK_COMMAND = "/back";
+
+    public final static LocalButton YES_BUTTON = new LocalButton("Да", ACCEPT_COMMAND);
+    public final static LocalButton NO_BUTTON = new LocalButton("Неа", DECLINE_COMMAND);
```
1. static запрещён
2. Срочно надо вспомнить русский язык

ASHGOLDOFFICIAL replied:
d1f4d6609c17ede56cf9fea6c1d3e20d3b3130cb

vpyzhyanov commented on src/main/java/ru/urfu/logics/mathmechbot/Constants.java:
```diff
@@ -0,0 +1,28 @@
+ */
+public final class Constants {
+    /**
+     * Приватный конструктор для класса-утилит.
```
в этом классе нет ни одной утилиты

ASHGOLDOFFICIAL replied:
169d35feaaacbbd1b6c524e36d428dfef39701ea

IDEA называла классы со статическими полями так, если не ошибаюсь.

vpyzhyanov commented on src/main/java/ru/urfu/logics/mathmechbot/Constants.java:
```diff
@@ -0,0 +1,28 @@
+
+    public final static LocalMessage TRY_AGAIN = new LocalMessageBuilder().text("Попробуйте снова.").build();
+    public final static LocalMessage ASK_FOR_REGISTRATION = new LocalMessageBuilder()
+            .text("Сперва нужно зарегистрироваться.").build();
```
вот посмотрите как вы усложнили код неуместным использование шаблона, вместо того, чтобы написать так:

`LocalMessage ASK_FOR_REGISTRATION = new LocalMessage("Сперва нужно зарегистрироваться.");`

DJettt replied:
Добавили конструктор в LocalMessage и исправили подобные моменты в коде.
da0d034721e24b3d6b7997626c843556d8cf2f94

vpyzhyanov commented on src/main/java/ru/urfu/logics/mathmechbot/MathMechBotCore.java:
```diff
@@ -0,0 +1,52 @@
+
+
+/**
+ * Логическое ядро бота, обрабатывающего сообщения Telegram-каналов на предмет упоминания студентов.<br/>
```
никаких связей с Telegram быть не должно

ASHGOLDOFFICIAL replied:
Наш бот на уровне идеи парсит Telegram-каналы. Парсить другие ресурсы изначально не предполагалось. Дождусь Артёма.

DJettt replied:
Все так. Бот будет парсить сообщение из Telegram-канала МатМеха, анализировать его и отправлять результат обработки в тот мессенджер, который укажет пользователь из доступных.

vpyzhyanov commented on src/main/java/ru/urfu/logics/mathmechbot/MathMechBotCore.java:
```diff
@@ -0,0 +1,52 @@
+
+        if (userOptional.isEmpty()) {
+            getStorage().getUsers().add(new User(request.id(), MathMechBotUserState.DEFAULT));
+            assert getStorage().getUsers().get(request.id()).isPresent();
```
1. почему вы не доверяете сами себе? В предыдущей строчке вы положили значение. Поэтому проверка никогда не сработает. Такая проверка используется, когда заранее неизвестно что может прийти на **вход** метода
2. Вы перепутали тест и основной код

ASHGOLDOFFICIAL replied:
99cf402ce1dc15aa991a50d42f11c3d470675ad6

assert был добавлен для IDEA, которая думала, что пользователь всё равно может отсутствовать.

DJettt replied:
Это было сделано буквально для уменьшения количества warning-ов и подчеркиваний со стороны IDEA Intellij.

ASHGOLDOFFICIAL replied:
Ну, фактически IDEA права. Нет никакой гарантии того, что никто не удалил объект после того, как мы положили объект в хранилище, и перед тем, как мы его оттуда вытащали. Нам помогает лишь то, что User'ов мы только создаём. Нигде не должно вызываться методов по их удалению. Но в любом случае, приведённое вами решение убирает потребность в assert.

vpyzhyanov replied:
научитесь создавать переменные и не делать одинаковые действия, тогда и лишних предупреждений не будет

vpyzhyanov commented on src/main/java/ru/urfu/logics/mathmechbot/MathMechBotCore.java:
```diff
@@ -0,0 +1,52 @@
+            getStorage().getUsers().add(new User(request.id(), MathMechBotUserState.DEFAULT));
+            assert getStorage().getUsers().get(request.id()).isPresent();
+        }
+        final User user = getStorage().getUsers().get(request.id()).get();
```
а может не будем дублировать код? И тем более не будем по несколько раз обращаться к хранилищу данных за одним и тем же?

```java
        UserStorage userStorage = getStorage().getUsers();
        final User user = userStorage.get(request.id())
                .orElseGet(() -> {
                    User newUser = new User(request.id(), MathMechBotUserState.DEFAULT);
                    userStorage.add(newUser);
                    return newUser;
                });
```

ASHGOLDOFFICIAL replied:
99cf402ce1dc15aa991a50d42f11c3d470675ad6

vpyzhyanov commented on src/main/java/ru/urfu/logics/mathmechbot/MathMechBotCore.java:
```diff
@@ -0,0 +1,52 @@
+     * Геттер поля storage.
+     * @return storage
+     */
+    public MathMechStorage getStorage() {
```
хм... инкапсуляция вышла погулять.... есть подозрение что необходимости в этом нет

ASHGOLDOFFICIAL replied:
Если вы про геттеры хранилищ, то я вам об этом говорил. Вы мне ещё показывали попытку сделать хранилище null-ом, что не вышло, так как поле -- final. В конце концов вы всё равно сказали, что нужно инкапсулировать, потому что это правильно.

Если вы про что-то другое, то я вас не понял.

Исправлять пока ничего не буду -- дождусь окончательного ответа по этому поводу.

vpyzhyanov replied:
обсудили на паре. Исправления не последовало( Метод должен быть удалён

vpyzhyanov commented on src/main/java/ru/urfu/logics/mathmechbot/MathMechBotCore.java:
```diff
@@ -0,0 +1,52 @@
+        final User user = getStorage().getUsers().get(request.id()).get();
+
+        currentState = user.currentState().stateInstance();
+        currentState.processMessage(this, request);
```
и очень подозрительно, что логика бота пошла в обработку состояния. Уже сейчас очевидно, что ООП тут и не пахнет, т.к. работа одного частного состояния требует общей логики. Это означает, что код сильно связан друг с другом. Например, чтобы изменить работу общей логики, вам придётся переписывать **каждое** состояние, которых может быть тысяча.

ASHGOLDOFFICIAL replied:
Состояние диктует то, как бот должен обрабатывать сообщение. Поэтому контекст передаёт обработку сообщения нужному состоянию. Не совсем понимаю, где состоянию требуется общая логика. Подожду Артёма.

Вообще, логика работы с состояниями будет переписана в третьем таске с введением FSM, если вы это согласуете, потому что пока я тоже не удовлетворён тем, как всё организовано (но не уверен, что тем же, чем и вы). В частности переходы в FSM будут диктовать, откуда в какое состояние можно перейти и что в процессе нужно сделать.

vpyzhyanov replied:
тоже обсудили на паре

vpyzhyanov replied:
Не забывайте описать результат договорённости

vpyzhyanov commented on src/main/java/ru/urfu/logics/mathmechbot/models/UserBuilder.java:
```diff
@@ -0,0 +1,31 @@
+     * @param id идентификатор пользователя.
+     * @param currentState состояние пользователя.
+     */
+    public UserBuilder(@NotNull Long id, @NotNull MathMechBotUserState currentState) {
```
а тут вообще смысла в билдере не было никогда. Он ничем от обычного конструктора не отличается. Все поля обязательные.

ASHGOLDOFFICIAL replied:
Изначально полей было 4, два необязательных (telegramId, discordId). Исправлять везде конструкторы было болезненным опытом, поэтому было решено создать builder. Когда в дальнейшем класс будет расширяться (например, полями isBanned или isAdmin), билдер позволит не изменять код везде и не плодить конструкторов.

DJettt replied:
Исправлено.
8789f6f2cf8a626bd4afecb7ce1c82ee90434e84

vpyzhyanov commented on src/main/java/ru/urfu/logics/mathmechbot/models/UserEntry.java:
```diff
@@ -0,0 +1,41 @@
+     *
+     * @return строка с содержимым объекта в удобном для пользователя виде.
+     */
+    public String toHumanReadable() {
```
а вот это очень хорошо! Правильно, что не стали использовать для этого toString(). Огромный лайк!

ASHGOLDOFFICIAL replied:
toString() уже определён рекордой. Плюс, toString() выводит все поля, что удобно для дебага, в то время как toHumanReadable() выводит только основную информацию для удобства чтения.

vpyzhyanov commented on src/main/java/ru/urfu/logics/mathmechbot/models/UserEntry.java:
```diff
@@ -0,0 +1,41 @@
+ * @param group номер группы того же человека;
+ * @param userId идентификатор модели User того, указывающий на того, кому пересылать упоминания этого человека.
+ */
+public record UserEntry(
```
очень интересно, как может случиться так, что у пользователя будет несколько таких объектов?

ASHGOLDOFFICIAL replied:
Описано в одной из следующих задач. Мы планируем позволить пользователю подписываться на упоминания нескольких лиц. Полезно старостам и кураторам.

vpyzhyanov commented on src/main/java/ru/urfu/logics/mathmechbot/models/UserEntryBuilder.java:
```diff
@@ -0,0 +1,142 @@
+/**
+ * Билдер для рекорда UserEntry.
+ */
+public final class UserEntryBuilder {
```
должен быть внутренним классом

ASHGOLDOFFICIAL replied:
Пока не вижу способов сделать это без static. Пока думать буду.

ASHGOLDOFFICIAL replied:
Ничего не придумал. Поэтому прошу вашего совета о том, как это сделать без static.

DJettt replied:
То же самое, что и с LocalMessageBuilder. Было решено оставить все как есть, чтобы не использовать static.

vpyzhyanov commented on src/main/java/ru/urfu/logics/mathmechbot/states/DefaultState.java:
```diff
@@ -0,0 +1,152 @@
+    private final static String DELETE_COMMAND = "/delete";
+
+    @Override
+    public void processMessage(@NotNull MathMechBotCore context, @NotNull Request request) {
```
Похоже ваша цель максимально запутать разработчиков. Ведь нет лучшего средства, чем называть вещи другими именами, например MathMechBotCore называть context.

ASHGOLDOFFICIAL replied:
context -- устоявшейся термин в паттерне State. Наименование шло согласно паттерну. Жду Артёма для обсуждения.

DJettt replied:
Переименовали context -> contextCore.
2bf244e07e4b2cf0da8596208798e7952f22f301

vpyzhyanov replied:
Лучше не стало. Если бездумно применять паттерн, то ничего полезного из него не выйдет

vpyzhyanov commented on src/main/java/ru/urfu/logics/mathmechbot/states/DefaultState.java:
```diff
@@ -0,0 +1,152 @@
+    private void registerCommandHandler(@NotNull MathMechBotCore context, @NotNull Request request) {
+        final Optional<UserEntry> userEntryOptional = context.getStorage().getUserEntries().get(request.id());
+        if (userEntryOptional.isEmpty()) {
+            context.getStorage().getUsers().changeUserState(request.id(), MathMechBotUserState.REGISTRATION_NAME);
```
ого, вот это да, то есть любая команда имеет полный доступ к данным абсолютно всех пользователей. Достаточно забыть сделать санитайзинг запроса и можно выполнить полное удаление данных)))

Конечно же так делать нельзя, как минимум пользователь должен работать только со своим состоянием. А если необходимо управлять чужим, то должен быть отдельный класс "хранилище данных", которое будет контролировать этот процесс

ASHGOLDOFFICIAL replied:
Я честно не понял, что не так. Во время регистрации просто смотрится, есть ли у данного пользователя хоть какая-то запись. Если нет, мы кидаем его в регистрацию. Я не вижу того, где бы пользователь работал не со своим состоянием. Артём, спасай.

DJettt replied:
Было решено перенести решение этой проблемы в task-3

vpyzhyanov commented on src/main/java/ru/urfu/logics/mathmechbot/states/DefaultState.java:
```diff
@@ -0,0 +1,152 @@
+        if (userEntryOptional.isPresent()) {
+            context.getStorage().getUsers().changeUserState(request.id(), MathMechBotUserState.DELETION_CONFIRMATION);
+
+            final LocalMessage msg = DeletionConfirmationState.INSTANCE.enterMessage(context, request);
```
static запрещён! Следовательно синглтон тоже

ASHGOLDOFFICIAL replied:
c120587db49c91c8a9584d404a80511aa171ebed

vpyzhyanov commented on src/main/java/ru/urfu/logics/mathmechbot/states/MathMechBotState.java:
```diff
@@ -0,0 +1,33 @@
+
+    /**
+     * <p>Вызывать этот метод, чтобы отправилось сообщение, которое должно отправлять при переходе в состояние.</p>
+     * TODO: сделать умнее.
```
что значит умнее? Пишите конкретно

ASHGOLDOFFICIAL replied:
Мне не нравится, что enterMessage() -- это целый метод, возвращающий в большинстве случаев статичный объект. Избавился от этого с помощью FSM, которую написал. Поэтому, когда будем мёржить её (надеюсь, в следующей задаче), эта проблема будет решена.

ASHGOLDOFFICIAL replied:
4f1f3a0176c94e76dabde18c2bccb516778859c4

vpyzhyanov commented on src/main/java/ru/urfu/logics/mathmechbot/states/deletion/DeletionConfirmationState.java:
```diff
@@ -0,0 +1,95 @@
+ * Состояние ожидания подтверждения удаления данных.
+ */
+public enum DeletionConfirmationState implements MathMechBotState {
+    INSTANCE;
```
нельзя. Тем более состояние синглтоном быть не может

ASHGOLDOFFICIAL replied:
c120587db49c91c8a9584d404a80511aa171ebed

vpyzhyanov commented on src/main/java/ru/urfu/logics/mathmechbot/states/deletion/DeletionConfirmationState.java:
```diff
@@ -0,0 +1,95 @@
+public enum DeletionConfirmationState implements MathMechBotState {
+    INSTANCE;
+
+    private final static Logger LOGGER = LoggerFactory.getLogger(DeletionConfirmationState.class);
```
нельзя

ASHGOLDOFFICIAL replied:
c120587db49c91c8a9584d404a80511aa171ebed

vpyzhyanov commented on src/main/java/ru/urfu/logics/mathmechbot/states/deletion/DeletionConfirmationState.java:
```diff
@@ -0,0 +1,95 @@
+        final Optional<UserEntry> userEntryOptional = context.getStorage().getUserEntries().get(request.id());
+        userEntryOptional.ifPresent(context.getStorage().getUserEntries()::delete);
+
+        context.getStorage().getUsers().changeUserState(request.id(), MathMechBotUserState.DEFAULT);
```
надо срочно учиться создавать переменные. Иначе вы уточните в дублируемом коде

ASHGOLDOFFICIAL replied:
688199873dd6cd7b5d5c5a372a83d747c653fc18

vpyzhyanov commented on src/main/java/ru/urfu/logics/mathmechbot/states/editing/EditingFullNameState.java:
```diff
@@ -0,0 +1,96 @@
+ * Состояние изменения ФИО.
+ */
+public enum EditingFullNameState implements MathMechBotState {
+    INSTANCE;
```
все остальные тоже

ASHGOLDOFFICIAL replied:
c120587db49c91c8a9584d404a80511aa171ebed

vpyzhyanov commented on src/main/java/ru/urfu/logics/mathmechbot/states/editing/EditingFullNameState.java:
```diff
@@ -0,0 +1,96 @@
+     * @return строка корректна или нет
+     */
+    public boolean validateFullName(String str) {
+        // TODO: Проверить более сложные имена, содержащие дефисы или несколько слов.
```
надо проверить и убрать комментарий

ASHGOLDOFFICIAL replied:
30aa462b317b4922cedffa04d2615d997357f3e9

vpyzhyanov commented on src/main/java/ru/urfu/logics/mathmechbot/states/editing/EditingFullNameState.java:
```diff
@@ -0,0 +1,96 @@
+        final List<String> strings = List.of(trimmedText.split("\\s+"));
+
+        context.getStorage().getUserEntries().changeUserEntrySurname(request.id(), strings.get(0));
+        context.getStorage().getUserEntries().changeUserEntryName(request.id(), strings.get(1));
```
срочно учимся создавать переменные

ASHGOLDOFFICIAL replied:
688199873dd6cd7b5d5c5a372a83d747c653fc18

vpyzhyanov commented on src/main/java/ru/urfu/storages/ArrayStorage.java:
```diff
@@ -0,0 +1,63 @@
+ * @param <I> тип идентификатора у хранимых объектов.
+ */
+public class ArrayStorage<T extends Identifiable<I>, I> implements Storage<T, I> {
+    private final ArrayList<T> array;
```
используйте интерфейсы

ASHGOLDOFFICIAL replied:
ArrayStorage -- это реализация Storage на массиве. Почему это должно быть интерфейсом?

ASHGOLDOFFICIAL replied:
4613e044bfcad80a6ad3d749ec82070bce1c72b4

Всё, я понял, что вы имели в виду. GitHub, который показывает сразу 4 строки ввёл меня в заблуждение.

vpyzhyanov replied:
Замечание пишется на конкретную строку, она самая нижняя. Верхние - приведены для контекста

vpyzhyanov commented on src/main/java/ru/urfu/logics/mathmechbot/storages/UserStorage.java:
```diff
@@ -0,0 +1,22 @@
+     * @param state новое текущее состояние.
+     */
+    default void changeUserState(@NotNull Long id, @NotNull MathMechBotUserState state) {
+        update(new UserBuilder(id, state).build());
```
некорректно так использовать дефолтные методы, т.к. по сути тут вы написали **реализацию**. Зачем тогда создавали интерфейс?

ASHGOLDOFFICIAL replied:
Эти реализации являются композицией методов, определённых в Storage. Реализация Storage влечёт за собой реализацию этих методов. Без реализации Storage эти методы ничего не делают. При этом тот факт, что ArrayStorage extends Storage обязывает эти методы реализовать.

ASHGOLDOFFICIAL replied:
Артём, оставлю тебе этот вопрос на решение.

ASHGOLDOFFICIAL replied:
А как, вообще, тогда корректно использовать дефолтные методы?

А интерфейс создавали для того, что бы реализовывать его с помощью ArrayStorage или иной реализации Storage, потому что повторюсь: данный код сам по себе ничего не делает. Если бы я переделал его в класс, он бы не запустился даже, так как требовал бы реализации используемых и неиспользуемых get, getAll, add, update, delete.

DJettt replied:
Нам интересно услышать Ваш ответ на наш вопрос, но в любом случае, это было изменено в этом коммите:
6fd51b66437d062d93fdfb446e3d660992c1c016

vpyzhyanov replied:
ответил на паре

vpyzhyanov commented on src/main/java/ru/urfu/logics/mathmechbot/storages/UserEntryStorage.java:
```diff
@@ -0,0 +1,107 @@
+     * @param id идентификатор записи в хранилище.
+     * @param strings ФИО
+     */
+    default void changeUserEntryNameSurname(@NotNull Long id, @NotNull List<String> strings) {
```
не надо путать интерфейс с реализацией

Сразу видно, что увлекаетесь спрингом. Нельзя бездумно переносить шаблоны из спринга на обычный код. ООП тут точно не будет.

ASHGOLDOFFICIAL replied:
Никто из нас со Spring'ом не работал. Поэтому не понимаю, о каких шаблоннах идёт речь.

Дополнительные методы в UserEntryStorage по редактированию отдельных полей были добавлены, так как я считаю, что за изменение полей должно отвечать хранилище. Ну, в случае с БД, мы данные редактируем в ней, а не вне её. Изменение выгруженных данных не влечёт изменения их в постоянном хранилище. Только если менять их напрямую в БД они будут меняться. Поэтому эти методы и были назначены Storage'ам, как часть их ответственности.

DJettt replied:
6fd51b66437d062d93fdfb446e3d660992c1c016
Изменено. Реализации методов были перенесены в соответствующие классы.

vpyzhyanov replied:
> Дополнительные методы в UserEntryStorage по редактированию отдельных полей были добавлены, так как я считаю, что за изменение полей должно отвечать хранилище. Ну, в случае с БД, мы данные редактируем в ней, а не вне её. Изменение выгруженных данных не влечёт изменения их в постоянном хранилище. Только если менять их напрямую в БД они будут меняться. Поэтому эти методы и были назначены Storage'ам, как часть их ответственности.

Я вам про Фому, а вы мне про Ерёму. Такими темпами вы точно ничему не научитесь. Сначала надо научиться слышать и читать. Потом научиться задавать вопросы. Далее можно будет научиться чему угодно. А пока вы не читаете замечания и пишете совершенно не связанные с ним вещи, ничего исправить и чему-то научиться точно не получится. Будете и дальше создавать совершенно бессмысленные классы вроде UserEntryArrayStorage, у которого реализация методов в **интерфейсе**, кто бы мог подумать.

vpyzhyanov commented on src/main/java/ru/urfu/storages/Identifiable.java:
```diff
@@ -0,0 +1,13 @@
+     * Идентификатор.
+     * @return идентификатор.
+     */
+    I id();
```
обычно методы получения данных начинаются с get, т.е. `getId()`

ASHGOLDOFFICIAL replied:
673483a0e3d19909e35722c885b84f79c7716f3f

vpyzhyanov commented on src/main/java/ru/urfu/storages/Storage.java:
```diff
@@ -0,0 +1,48 @@
+     *
+     * @param member элемент, который добавляем.
+     */
+    void add(T member) throws IllegalArgumentException;
```
нужно в JavaDoc описать это исключение и в каком случае оно произойдёт

ASHGOLDOFFICIAL replied:
9e14d4151ee8f955788690923204ed17a97400dd

vpyzhyanov commented on src/test/java/ru/urfu/logics/DummyBot.java:
```diff
@@ -0,0 +1,38 @@
+     * Конструктор.
+     */
+    public DummyBot() {
+        outcomingMessageList = new Stack<>();
```
стек? Вы уверены?

ASHGOLDOFFICIAL replied:
24392b00ab844dd63a42d2c842fc5e65db62ec1e

Изначально думал, что раз мы в большинстве случаев достаёт только последнее сообщение, можно использовать стек. Потом я об этом забыл. И фактически мы не только последние сообщения доставали. Исправил, в общем-то.

vpyzhyanov commented on src/test/java/ru/urfu/logics/EchoBotCoreTest.java:
```diff
@@ -0,0 +1,92 @@
+ * Тесты для класса EchoBotCore.
+ */
+@DisplayName("Логика эхо бота")
+final class EchoBotCoreTest {
```
зачем? Просто удалите это логику, она не нужна

ASHGOLDOFFICIAL replied:
c7bb4747ad69050adda85e75ef6650e730fe219f

vpyzhyanov commented on src/test/java/ru/urfu/logics/mathmechbot/DeleteCommandTest.java:
```diff
@@ -0,0 +1,177 @@
+     * </ol>
+     */
+    @Test
+    @DisplayName("Нажата кнопка 'Неа' во время подтверждения удаления данных")
```
ещё и в тесте зафиксировать незнание русского языка

ASHGOLDOFFICIAL replied:
1cb7e79d2f3a53494b704e22488489648cff6eeb

Неформальная форма была умышленно, так как наша целевая аудитория -- студенты. Использование неформальных форм делает бота хотя сколько-то приближенным к целевой аудитории.

vpyzhyanov replied:
или показывает, что разработчики безграмотные и приложение их не серьёзное, а шуточное, т.е. ничего полезного не делает. Не забывайте с разных сторон посмотреть на одни и те же вещи

vpyzhyanov commented on src/test/java/ru/urfu/logics/mathmechbot/TestConstants.java:
```diff
@@ -0,0 +1,86 @@
+    public final static LocalMessage ACCEPT_MESSAGE = new LocalMessageBuilder().text(ACCEPT_COMMAND).build();
+
+    public final static String DECLINE_COMMAND = "/no";
+    public final static LocalMessage DECLINE_MESSAGE = new LocalMessageBuilder().text(DECLINE_COMMAND).build();
```
запрещено

ASHGOLDOFFICIAL replied:
1fdadb9634c4a9202eccd4b01eac4bc3e28bbe6c

vpyzhyanov commented on src/test/java/ru/urfu/logics/mathmechbot/TestUtils.java:
```diff
@@ -0,0 +1,77 @@
+     * @param men       группа в формате МЕН.
+     */
+    public void registerUser(long id, String fullName, int year, String specialty, int group, String men) {
+        final ArrayList<LocalMessage> messages = new ArrayList<>(List.of(
```
1. Используйте интерфейсы
2. Пощадите оперативную память и не создавайте 2 абсолютно одинаковые коллекции

ASHGOLDOFFICIAL replied:
4426662008b88ecac3edee1c8725a254d20c4200

vpyzhyanov commented on src/test/java/ru/urfu/logics/mathmechbot/editing/EditingTest.java:
```diff
@@ -0,0 +1,274 @@
+    private MathMechBotCore logic;
+    private DummyBot bot;
+
+    User currentUser;
```
отсутствует инкапсуляция

ASHGOLDOFFICIAL replied:
612b8aef67d77a8f23e2d05cb43bd2e89d5d69a5

Артём, проверь, нигде ли я не забыл добавить private.

vpyzhyanov commented on src/test/java/ru/urfu/logics/mathmechbot/editing/EditingTest.java:
```diff
@@ -0,0 +1,274 @@
+                bot.getOutcomingMessageList().getFirst());
+        Assertions.assertEquals(
+                new UserBuilder(0L, MathMechBotUserState.EDITING_CHOOSE).build(),
+                storage.getUsers().get(0L).orElseThrow());
```
@по хорошему тут вы не должны иметь доступ к хранилищу данных. Значит в инкапсуляции жирная дыра

Состояния - это детали реализации, о которых вообще никто знать не должен, даже тест

ASHGOLDOFFICIAL replied:
Слишком большая проблема, чтобы исправлять одному. Сперва проконсультируюсь с Артёмом.

vpyzhyanov commented on src/test/java/ru/urfu/logics/mathmechbot/editing/EditingTest.java:
```diff
@@ -0,0 +1,274 @@
+                .text(TestConstants.EDITING_FULL_NAME_COMMAND)
+                .build()));
+        Assertions.assertEquals(
+                new UserBuilder(0L, MathMechBotUserState.EDITING_FULL_NAME).build(),
```
я про это рассказывал на лекции. Самый первый антипример. Вы не должны использовать внутренние классы для проверки. Тут ситуация чуть сложнее, но суть та же

ASHGOLDOFFICIAL replied:
Артём, надо исправить тесты.

DJettt replied:
65c1c9ee8da3e9a2d02fbef6cee7633c7d74d310

vpyzhyanov reviewed (changes requested):
Вы почему отправляете решение, которое не удовлетворяет элементарным требованиям? В этот раз я проверил, но в следующий раз даже проверять не буду

ASHGOLDOFFICIAL committed:
Убрал ненужный метод в Main, убрал start у Bot.

ASHGOLDOFFICIAL committed:
Вынес создание хранилищ внутрь конструктора MathMechStorage

ASHGOLDOFFICIAL committed:
Merge remote-tracking branch 'refs/remotes/origin/task-2' into task-2-fixes

ASHGOLDOFFICIAL committed:
Массовое удаление static из основного кода. Тесты не трогал. Добавил отсутствующие "case null".

ASHGOLDOFFICIAL committed:
Убрал "пасхалку" из DiscordBot.

ASHGOLDOFFICIAL committed:
Чистка DiscordBot.

ASHGOLDOFFICIAL committed:
Чистка TelegramBot.

ASHGOLDOFFICIAL committed:
Заменил long у id в Request на обёртку Long, повесил NotNull.

ASHGOLDOFFICIAL committed:
"Неа" -> "Нет".

ASHGOLDOFFICIAL committed:
Упростил MathMechBotCore#processMessage.

ASHGOLDOFFICIAL committed:
Identifiable#id -> Identifiable#getId

ASHGOLDOFFICIAL committed:
Пояснение исключения у Storage#add.

ASHGOLDOFFICIAL committed:
Удалил EchoBotCore.

ASHGOLDOFFICIAL committed:
"Неа" -> "Нет"

ASHGOLDOFFICIAL committed:
Массовое удаление static в тестах.

ASHGOLDOFFICIAL committed:
Сделал приватными поля в тестах, которые до этого были package-private.

ASHGOLDOFFICIAL committed:
Убрал singleton-enum'ы.

ASHGOLDOFFICIAL committed:
Убрал незамеченный синглтон-енам. Повыделял хранилища в отдельные переменные.

ASHGOLDOFFICIAL committed:
Stack -> ArrayList в DummyBot.

ASHGOLDOFFICIAL committed:
Заменил ArratList<>(List.of(...)) -> List.of(...)

ASHGOLDOFFICIAL committed:
Использую интерфейс в ArrayStorage.

ASHGOLDOFFICIAL committed:
Убрал неиспользуемый метод в UserEntryStorage.

ASHGOLDOFFICIAL committed:
Вынес версию JDA в переменную.

DJettt committed:
context -> contextCore

DJettt committed:
фикс оформления

ASHGOLDOFFICIAL committed:
Merge pull request #20 from DJettt/task-2-fixes

Исправления ко второй задаче.

ASHGOLDOFFICIAL committed:
Разъяснил TODO.

ASHGOLDOFFICIAL committed:
Merge remote-tracking branch 'origin/task-2-fixes' into task-2-fixes

DJettt committed:
Добавил конструктор по строке в LocalMessage, заменил LocalMessageBuilder на этот конструктор везде где устанавливается только поле text.

ASHGOLDOFFICIAL committed:
Поддержка дефисов в ФИО. Соответствующие тесты.

DJettt committed:
Добавил кнопку "Отменить регистрацию" в самом начале регистрации.

DJettt committed:
Убрал UserBuilder, теперь везде используются конструкторы User.

ASHGOLDOFFICIAL committed:
Раскидал константные сообщения по классам.

ASHGOLDOFFICIAL committed:
Merge remote-tracking branch 'refs/remotes/origin/task-2-fixes-pribylov' into task-2-fixes

# Conflicts:
#	src/main/java/ru/urfu/logics/mathmechbot/Constants.java
#	src/main/java/ru/urfu/logics/mathmechbot/states/editing/EditingSpecialityState.java

ASHGOLDOFFICIAL committed:
Упрощение тестов.

ASHGOLDOFFICIAL committed:
Обновил JavaDoc'и в тестах удаления.

ASHGOLDOFFICIAL committed:
Исправления JavaDoc'ов.

ASHGOLDOFFICIAL committed:
Красота в тестах.

DJettt committed:
Merge pull request #22 from DJettt/task-2-fixes-pribylov

Task 2 fixes pribylov

ASHGOLDOFFICIAL committed:
Merge pull request #21 from DJettt/task-2-fixes

Исправления ко второй задаче.

ASHGOLDOFFICIAL committed:
Стиль тестов.

DJettt committed:
Оформил тесты. Теперь в них нет прямого обращения к storage.

DJettt committed:
Вынес тела методов из интерфейсов UserStorage и UserEntryStorage в соответствующие им объекты.

DJettt committed:
Tests fixed

DJettt committed:
Tests fixed 2

DJettt committed:
Merge pull request #23 from DJettt/task-2-fixes-pribylov

Task 2 fixes pribylov

DJettt committed:
Убрал Request

DJettt committed:
Merge pull request #24 from DJettt/task-2-fixes-pribylov

Убрал Request

ASHGOLDOFFICIAL committed:
Merge remote-tracking branch 'refs/remotes/origin/task-2' into task-2-fixes

# Conflicts:
#	src/test/java/ru/urfu/logics/mathmechbot/RegistrationTest.java

ASHGOLDOFFICIAL committed:
bug fix: переходы FullName -> Group -> FullName -> Group ломали бота
На этапе FullName создавалась UserEntry, но при возврате из Group она не удалялась

ASHGOLDOFFICIAL committed:
Merge pull request #25 from DJettt/task-2-fixes

Исправления ко второй задаче (3).

ASHGOLDOFFICIAL committed:
Оформление JavaDoc'ов.

ASHGOLDOFFICIAL committed:
Фикс стиля.

ASHGOLDOFFICIAL committed:
Merge pull request #29 from DJettt/task-2-fixes

Исправления ко второй задаче (4).

vpyzhyanov reviewed (approved):
можете мёрджить
